src/jquery/attributes.js

```diff
@@ -1,9 +1,116 @@
 import { migratePatchFunc, migrateWarn } from "../main.js";
+import { jQueryVersionSince } from "../compareVersions.js";
 
 var oldRemoveAttr = jQuery.fn.removeAttr,
+	oldJQueryAttr = jQuery.attr,
 	oldToggleClass = jQuery.fn.toggleClass,
-	rbooleans = /^(?:checked|selected|async|autofocus|autoplay|controls|defer|disabled|hidden|ismap|loop|multiple|open|readonly|required|scoped)$/i,
-	rmatchNonSpace = /\S+/g;
+	booleans = "checked|selected|async|autofocus|autoplay|controls|defer|" +
+		"disabled|hidden|ismap|loop|multiple|open|readonly|required|scoped",
+	rbooleans = new RegExp( "^(?:" + booleans + ")$", "i" ),
+	rmatchNonSpace = /\S+/g,
+
+	// Some formerly boolean attributes gained new values with special meaning.
+	// Skip the old boolean attr logic for those values.
+	extraBoolAttrValues = {
+		hidden: [ "until-found" ]
+	};
+
+// HTML boolean attributes have special behavior:
+// we consider the lowercase name to be the only valid value, so
+// getting (if the attribute is present) normalizes to that, as does
+// setting to any non-`false` value (and setting to `false` removes the attribute).
+// See https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes
+jQuery.each( booleans.split( "|" ), function( _i, name ) {
+	var origAttrHooks = jQuery.attrHooks[ name ] || {};
+	jQuery.attrHooks[ name ] = {
+		get: origAttrHooks.get || function( elem ) {
+			var attrValue;
+
+			if ( jQuery.migrateIsPatchEnabled( "boolean-attributes" ) ) {
+				attrValue = elem.getAttribute( name );
+
+				if ( attrValue !== name && attrValue != null &&
+					( extraBoolAttrValues[ name ] || [] ).indexOf( attrValue ) === -1
+				) {
+					migrateWarn( "boolean-attributes",
+						"Boolean attribute '" + name +
+							"' value is different from its lowercased name" );
+
+					// jQuery <4 attr hooks setup is complex: there are attr
+					// hooks, bool hooks and selector attr handles. Only
+					// implement the logic in jQuery >=4 where it's missing
+					// and there are only attr hooks.
+					if ( jQueryVersionSince( "4.0.0" ) ) {
+						return name.toLowerCase();
+					}
+					return null;
```

Comment:
This works in jQuery 3.x because - while `boolHooks` being a fallback for boolean attributes if `attrHooks[ name ]` didn't exist would not get picked up here - `boolHooks` only handle a setter and we're in a getter here. The getter - if the `get` attr hook is missing or returns `null` - is handled via a call to `jQuery.find.attr( elem, name )` which is `Sizzle.attr( elem, name )`, that defers to `Sizzle.selectors.attrHandle`, exposed to jQuery via `jQuery.expr.attrHandle` and that, in turn, is set for boolean attrs by jQuery.